### PR TITLE
netmap support

### PR DESCRIPTION
**Pull request by gureedo**

### CODE REVIEW

**Comment by gureedo** [comment]
:up:

**Comment by inliniac** [comment]
Looks good to me. Won't be able to test it though.

**Comment by cheffo** [comment]
@ndenev ? :)

**Comment by ndenev** [comment]
I presume this was developed and tested on Linux?
I think I'll be able to test this on FreeBSD 10.1 in the next few days.

**Comment by gureedo** [comment]
`I presume this was developed and tested on Linux?`
yep!

**Comment by ndenev** [comment]
Compilation on FreeBSD 10.1 seems to fail with the following:
```
source-netmap.c:678:24: warning: initializing 'unsigned char *' with an expression of type 'char *' converts between pointers to integer types with different sign [-Wpointer-sign]
        unsigned char *slot_data = NETMAP_BUF(ring, slot->buf_idx);
                       ^           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
source-netmap.c:798:43: error: use of undeclared identifier 'POLLRDHUP'
            if (fds[i].revents & (POLLHUP|POLLRDHUP|POLLERR|POLLNVAL)) {
                                          ^
1 warning and 1 error generated.
```

Looks like POLLRDHUP does not exist on FreeBSD, and the following patch helps this(there are still a few warnings from clang):
```
diff --git a/src/source-netmap.c b/src/source-netmap.c
index e99adf4..ed94b98 100644
--- a/src/source-netmap.c
+++ b/src/source-netmap.c
@@ -129,6 +129,12 @@ TmEcode NoNetmapSupportExit(ThreadVars *tv, void *initdata, void **data)
 
 #define POLL_TIMEOUT 100
 
+#if defined(__linux__)
+#define POLL_EVENTS POLLHUP|POLLRDHUP|POLLERR|POLLNVAL
+#else
+#define POLL_EVENTS POLLHUP|POLLERR|POLLNVAL
+#endif
+
 enum {
     NETMAP_OK,
     NETMAP_FAILURE,
@@ -795,7 +801,7 @@ static TmEcode ReceiveNetmapLoop(ThreadVars *tv, void *data, void *slot)
         }
 
         for (int i = 0; i < rings_count; i++) {
-            if (fds[i].revents & (POLLHUP|POLLRDHUP|POLLERR|POLLNVAL)) {
+            if (fds[i].revents & (POLL_EVENTS)) {
                 if (fds[i].revents & POLLERR) {
                     SCLogError(SC_ERR_NETMAP_READ,
                                "Error reading data from iface '%s': (%d" PRIu32 ") %s",
```

**Comment by gureedo** [comment]
@ndenev fixed

**Comment by ndenev** [comment]
Thanks, I'm recompiling now, and will let you know.
P.S.: Is it going to be possible to use copy-mode with a single interface? That is use suricata to bridge between the network interface and host rings?

**Comment by ndenev** [comment]
I had to enable promisc mode, but now I'm capturing traffic and I see entries in eve.json (this is on a port connected to a mirror port on my switch)

**Comment by gureedo** [comment]
Yes, you can pass packets with single interface (simple rx->tx forwarding).
Current implementation does not allow to communicate with host stack.

**Comment by ndenev** [comment]
Yes, I just saw that.
So, looks like won't be very hard to basically open with NR_REG_SW_NIC in case output if name is the same of input interface. I'll see if I'll have time to do that.
But in the mean time this seems to work nicely as it is, and I noticed zero dropped packets at 100Mbit/s on my Alix APU, whereas before, using pcap (even with increased pcap buffers, and bpf zero copy enabled) I was seeing some drops at higher speeds.

**Comment by gureedo** [comment]
I think there must be special semantics for communication with host stack.
Using one interface in ids mode gives ability to handle 10gb traffic (connect tx to input device and rx to another output device, i'm talking about optic connection) and save one network port :)

**Comment by gureedo** [comment]
:up:

**Comment by ndenev** [comment]
FWIW, this has been working perfectly fine here for the past ~2 weeks.

File: .gitignore [hunk 1]
```diff
@@ -36,3 +36,4 @@ src/suricata
 stamp-h1
 src/build-info.h
 qa/log/
+.idea
```
Comment (inliniac) [inline .gitignore]: Can you remove this from the PR? It's unrelated to netmap

**Comment by inliniac** [comment]
Can you rebase the code against the current master and do a new pull request?

**Comment by pavel-odintsov** [comment]
Will be fine to add drivers for popular NICs for netmap on Linux, I have prepared it for ixgbe driver https://github.com/pavel-odintsov/ixgbe-linux-netmap

**Pull request closed without merging**